src/client/Microsoft.Identity.Client/ManagedIdentity/ManagedIdentityClient.cs

```diff
@@ -47,5 +47,41 @@ private static AbstractManagedIdentity SelectManagedIdentitySource(RequestContex
                 AzureArcManagedIdentitySource.TryCreate(requestContext) ??
                 new ImdsManagedIdentitySource(requestContext);
         }
+
+        internal static ManagedIdentitySource GetManagedIdentitySource()
+        {
+            string identityEndpoint = EnvironmentVariables.IdentityEndpoint;
+            string identityHeader = EnvironmentVariables.IdentityHeader;
+            string identityServerThumbprint = EnvironmentVariables.IdentityServerThumbprint;
+            string msiSecret = EnvironmentVariables.IdentityHeader;
+            string msiEndpoint = EnvironmentVariables.MsiEndpoint;
+            string imdsEndpoint = EnvironmentVariables.ImdsEndpoint;
+            string podIdentityEndpoint = EnvironmentVariables.PodIdentityEndpoint;
+
+            if (!string.IsNullOrEmpty(identityEndpoint) && !string.IsNullOrEmpty(identityHeader) && !string.IsNullOrEmpty(identityServerThumbprint))
+            {
+                return ManagedIdentitySource.ServiceFabric;
+            }
+            else if (!string.IsNullOrEmpty(identityEndpoint) && !string.IsNullOrEmpty(identityHeader))
+            {
+                return ManagedIdentitySource.AppService;
+            }
+            else if (!string.IsNullOrEmpty(msiEndpoint))
+            {
+                return ManagedIdentitySource.CloudShell;
+            }
+            else if (!string.IsNullOrEmpty(identityEndpoint) && !string.IsNullOrEmpty(imdsEndpoint))
+            {
+                return ManagedIdentitySource.AzureArc;
+            }
+            else if (!string.IsNullOrEmpty(podIdentityEndpoint))
+            {
+                return ManagedIdentitySource.Imds;
+            }
+            else
+            {
+                return ManagedIdentitySource.DefaultToImds;
```

Comment:
I don't understand the difference between IMDS and DefaultToImds. It's just IMDS. Maybe it's just a naming problem - is the Pod releated to Kubernetes? In which case maybe call it KubernetesImds? 

It does not see to be as per spec: https://identitydivision.visualstudio.com/DevEx/_git/AuthLibrariesApiReview?version=GBdev&path=/ManagedIdentity/FacilitateImdsProbing.md&_a=preview&anchor=msal-implementations

Reply:
Ideally we should find this environment variable POD_IDENTITY on the VM. If that is the case, it definitely is a VM with managed identity configured. But since this environment variable is generally not found on the VM, we Default to VM.  I don't think I have seen the POD_IDENTITY environment variable ever during my testing and I am not sure what is the probability of having this set. CC: @rayluo

Reply:
I am not aware of the POD_IDENTITY env var either. Out of curiosity, is it defined in some docs, or was it backported from Azure Identity .Net's implementation?

Regardless, I suppose you may keep both the existing `Imds` logic and the new `DefaultToImds` logic. In such a case, the new enum name "*DefaultTo*Imds" actually plays well with your existing "Imds" name, because their different names hint their subtle differences.

Aside: Personally, I would choose the name "...VM" rather than "...Imds". VM is a kind of runtime environment, [IMDS is a service/endpoint available in VM](https://learn.microsoft.com/en-us/azure/virtual-machines/instance-metadata-service?tabs=linux). Many other Managed Identity environments use an IMDS-like endpoint, too.

Reply:
@rayluo, this was backported from Azire .Net's implementation.

I agree VM would have been a better choice but it is already there.

Reply:
KubernetesImds?

Reply:
That is a good point. These should be consistent across MSALs for telemetry. But the DefaultToImds is only sent back as response to GetManagedIdentitySource and not used for other scenarios.

Reply:
@rayluo, using the existing enum IMDS. Do you see any issues with it?

Reply:
Discussed offline